internal/clusterconfig/clusterconfig.go

```diff
@@ -94,6 +95,40 @@ func EvaluateClusterConfiguration(ctx context.Context, k8sClient client.Client)
 	return flavour.clusterConfiguration()
 }
 
+// ErrUnknownProvider indicates that provider is not known
+type ErrUnknownProvider struct{}
+
+func (e ErrUnknownProvider) Error() string {
+	return "unknown provider"
+}
+
+// GetClusterProvider is a small hack that tries to determine the
+// hyperscaler based on the first provider node.
+func GetClusterProvider(ctx context.Context, k8sclient client.Client) (string, error) {
+	nodes := corev1.NodeList{}
+	err := k8sclient.List(ctx, &nodes)
+	if err != nil {
+		return "", err
+	}
+	// if node list is empty, treat as unknown provider
+	if nodes.Items == nil {
+		return "", ErrUnknownProvider{}
+	}
+
+	// get 1st node since all nodes usually are backed by the same provider
+	n := nodes.Items[0]
+	provider := n.Spec.ProviderID
+	switch {
+	case strings.HasPrefix(provider, "aws://"):
+		return "aws", nil
+	case strings.HasPrefix(provider, "gce://"):
+		return "gce", nil
+	case strings.HasPrefix(provider, "openstack://"):
+		return "openstack", nil
```

Comment:
Hmm, generally I'm not really sure about this implementation, as it returns an error for not known providers. This way we need to specify ALL our cluster providers here for it not to return an error.
I would not expect it to error in this way, as the only information we only need here is whether it's AWS or not, and not being AWS is not an error, just a information

Reply:
These kinds of code should not be in the controller to begin with. I expect that it's up to the implementer to decide whether they want to handle an error or not. I rather wanted to have the error exposed, then handled as a return variable for better visibility in the logs. For now you can remove other cases besides AWS, though I think having it more flexible makes more sense to me.